### PR TITLE
Change service specfic env var overrides

### PR DESCRIPTION
**Pull request by rudderfree**
Change service specific environemnt overrides to be able to
use _ instead of - eg: use
support_notifications_Smtp_Subject instead of
support-notifications_Smtp_Subject

Issue no.: #1519

Signed-off-by: Amit Kumar Chaudhary <a.rudder.free@gmail.com>

### CODE REVIEW

**Comment by rudderfree** [comment]
Not sure what is going on here, error from the Jenkins logs

17:30:47 # github.com/edgexfoundry/edgex-go/internal/seed/config [github.com/edgexfoundry/edgex-go/internal/seed/config.test]
17:30:47 internal/seed/config/environment.go:55: undefined: strings.ReplaceAll

However, it is building on my machine. Log from my machine -

ok  	github.com/edgexfoundry/edgex-go/internal/seed/config	0.013s	coverage: 8.3% of statements

Probably my build environment is lagging behind in versions of something, Go(?). Will take a look at it.

**Comment by rudderfree** [comment]
Closing temporarily, will put it back up soon after finding the problem.

**Comment by rudderfree** [comment]
Recheck

**Comment by rudderfree** [comment]
While testing this change, I tried setting password from notification service to empty
[support_notifications_Smtp_Password=""], But it didn't override it. The following line appears to be making the difference. 

	`	if len(kv) == 2 && len(kv[0]) > 0 && len(kv[1]) > 0 {`

[Link to the above line](https://github.com/rudderfree/edgex-go/blob/a224da4f06dcbe362e752b2917792d43d17afe6e/internal/seed/config/environment.go#L46)

kv[1] should be >= 0 for the override to happen, but I am not sure about the overall design, so I was just wondering can the change be made or not.

**Comment by rudderfree** [comment]
> Don't believe this change set will have the desired effect.

I see what you mean. This being service specific change, it does override correctly. the global key part just passes through. I have tested service specific part multiple times, the issue talks about only service specific change.

> 
> Code in NewEnvironment() that reads and stores environment variables (and transforms embedded underscores to periods) does not have related conversion of hyphens. 

I was expecting something to change in this line too, when I first looked at it, but since the service-specific override vars will not have any hyphens (defined as such), the conversion is not required. But for those which do have hyphens will retain their hyphens, those would be global keys.

> Not sure how a hyphenated service name would match.

They do because the hyphenated ones are not getting compared with non hyphenated ones, non-hyphenated ones first get the service name hyphens converted to dots, which is what those (service name part) are getting compared against.

		`case strings.HasPrefix(k, serviceName):`

> Further, I'm not certain the modified service name would be found in the toml.Tree (which is created from configuration.toml) for a hyphenated service.

From the code, It seems service name is not playing a part when setting these variables in the tree. 
			`tree.Set(strings.TrimPrefix(k, serviceName), v)`

The above two lines already existed and I didn't see a reason to doubt them, More so because registry was updating as expected.

> 
> Recommend adding one or more tests to environment_test.go to cover the case(s) where service name contains a hyphen.

Working on it.

> 
> (You also need to rebase/squash your commit chain into a single commit.)

Will do.

File: internal/seed/config/environment.go [hunk 1]
```diff
@@ -52,6 +52,7 @@ func NewEnvironment() *environment {
 
 // OverrideFromEnvironment method replaces values in the toml.Tree for matching environment variable keys.
 func (e *environment) OverrideFromEnvironment(serviceName string, tree *toml.Tree) *toml.Tree {
+	serviceName = strings.Replace(serviceName, "-", ".", -1)
```
[the remark this answers is not in the document]
Comment (rudderfree) [reply internal/seed/config/environment.go]: Good point, will make the change.

[the remark this answers is not in the document]
Comment (rudderfree) [reply internal/seed/config/environment.go]: Done.

**Comment by rudderfree** [comment]
> > Recommend adding one or more tests to environment_test.go to cover the case(s) where service name contains a hyphen.
> 
> Working on it.

Added a case for hyphenated service name. As I better understand the global keys now, I think there is never going to be a case where hyphen will be found in them, but if does makes it way through a change in convention/design, it will get updated as expected. Added one for service name with underscore as well.

**Comment by rudderfree** [comment]
@michaelestrin  pushed the changes, please review when you get a chance, thanks!

File: internal/seed/config/environment_test.go [hunk 2]
```diff
@@ -62,8 +62,9 @@ func TestKeyMatchOverwritesValue(t *testing.T) {
 		{"generic sub", sub + "." + subKey, sub + "." + subKey, envValue, serviceName, envValue},
 		{"service root", rootKey, serviceName + "." + rootKey, envValue, serviceName, envValue},
 		{"service sub", sub + "." + subKey, serviceName + "." + sub + "." + subKey, envValue, serviceName, envValue},
+		{"service hyphen", rootKey, rootKey, envValue, "service-Name", envValue},
+		{"service underscore", rootKey, rootKey, envValue, "service_Name", envValue},
```
Comment (michaelestrin) [inline internal/seed/config/environment_test.go]: These tests don't exercise the variants they describe.  Note the other preexisting service tests ("service root" and "service sub") add the serviceName in the tests.envKey value to the rootKey and sub.subKey respectively.

There are two types of overrides -- service-specific (where the serviceName is part of the environment key) and global (where it is not).  

Your two tests succeed in spite of your serviceName variations -- they match based on rootKey = rootKey (the global version).  They need to be updated for the service-specific variation (by adding the service name variation to the test's envKey).  For example:

`{"service hyphen", rootKey, "service-Name" + "." + rootKey, envValue, "service-Name", envValue},`

You'll then need to refactor your code so the resulting test does not fail.

Also, to be consistent with the existing test implementation each variation of serviceName should have two separate test cases -- one for a root key match and one for a sub key match.

Comment (rudderfree) [reply internal/seed/config/environment_test.go]: I agree, added the above test case and others too. Refactored the code as well.

From the issue description, I had gathered that we are ditching the hyphens and moving to only use the underscores, but with the context from discussion above, the description can be easily re-read to extend the naming convention to include underscores also. So this refactor includes it along with the test cases. 

Please review when you get a chance, thanks!

File: internal/seed/config/environment_test.go [hunk 3]
```diff
@@ -62,8 +62,11 @@ func TestKeyMatchOverwritesValue(t *testing.T) {
 		{"generic sub", sub + "." + subKey, sub + "." + subKey, envValue, serviceName, envValue},
 		{"service root", rootKey, serviceName + "." + rootKey, envValue, serviceName, envValue},
 		{"service sub", sub + "." + subKey, serviceName + "." + sub + "." + subKey, envValue, serviceName, envValue},
+		{"service hyphen", rootKey, "service-Name" + "." + rootKey, envValue, "service-Name", envValue},
+		{"service hyphen sub", sub + "." + subKey, "service-Name" + "." + sub + "." + subKey, envValue, "service-Name", envValue},
+		{"service underscore", rootKey, "service_Name" + "." + rootKey, envValue, "service-Name", envValue},
+		{"service underscore sub", sub + "." + subKey, "service_Name" + "." + sub + "." + subKey, envValue, "service-Name", envValue},
```
Comment (michaelestrin) [inline internal/seed/config/environment_test.go]: Replace magic values (i.e. "service-Name" and "service_Name") with constants.

**Comment by rudderfree** [comment]
I didn't get a chance to work on it yet, planning to get to it over the weekend, Thx!

File: internal/seed/config/environment.go [hunk 4]
```diff
@@ -44,15 +44,15 @@ func NewEnvironment() *environment {
 	for _, env := range osEnv {
 		kv := strings.Split(env, "=")
 		if len(kv) == 2 && len(kv[0]) > 0 && len(kv[1]) > 0 {
-			e.env[strings.Replace(kv[0], "_", ".", -1)] = kv[1]
+			e.env[strings.Replace(strings.Replace(kv[0], "-", ".", -1), "_", ".", -1)] = kv[1]
```
[the remark this answers is not in the document]
Comment (rudderfree) [reply internal/seed/config/environment.go]: SonarCloud is assessing low score for repeated lines, I am contemplating changing to the code below.

```
func NewEnvironment() *environment {
        sep := []string{"-", "_"}
...
...
                        for _, s := range sep {
                                kv[0] = strings.Replace(kv[0], s, ".", -1)
                        }
                        e.env[kv[0]] = kv[1]
```

**Comment by michaelestrin** [comment]
> SonarCloud is assessing low score for repeated lines

Sonar is complaining about duplication in the 2.9K lines added since it was last run.  I don't see it complaining about these specific lines that you've added.  This is a configuration issue with Sonar and does not result from the changes you've made.

@tsconn23 Who is responsible for configuration of Sonar quality gates?  Any idea why this commit is driving a Sonar failure (this commit didn't add 2.9K additional lines of code) and any way to get the Sonar failure addressed?

**Comment by tsconn23** [comment]
recheck

**Comment by rudderfree** [comment]
@tsconn23 @michaelestrin, Thanks for the merge and closing.

**Pull request merged**